### ISSUE TITLE
Dynamic Data Support for Select

### ISSUE THREAD

**Issue by dev-cprice** [cerner/terra-core#2369]
# Issue Description

## TLDR
As it stands, the searchable `terra-form-select` API does not handle dynamic data well. 

## Reasoning

Let's say that I want to implement something similar to the following (which uses [select2](https://github.com/select2/select2) under the hood):

![2019-04-29 14 48 47](https://user-images.githubusercontent.com/34115417/56922540-09926480-6a8e-11e9-99da-579721708201.gif)

The API here is rather simple:
- store a list of selected values, both the `id` and displayable `text`
- display list of selected values as de-selectable `tags` (similar to what the current `terra-form-select` `multiple` and `tag` variants do)
- on search state (e.g. typing) lazy load data from some source (e.g. remote api, database, etc)
  - while loading, display a fallback UI, in this case: `"Searching..."`
  - on load, display the loaded list, sans any previous selections since they already appear as tags
- optionally: have a search query length threshold, in this case `2` chars
- optionally: have a max selection count that prevents the user from selecting more than the amount supplied

### Current Limitations

The current `terra-form-select` cannot perform the above for multiple reasons:

1. The selected items displayable text is coupled to the `<Select />`'s children `<Select.Option />`s.
    * The selected items display as as `<Tag />`s which will fallback to displaying their `value` (generally a GUID, or some human unreadable thing) if the option isn't a current child (See: [SelectUtil.valueDisplay](https://github.com/cerner/terra-core/blob/master/packages/terra-form-select/src/_SelectUtil.js#L84-L87) and [MenuUtil.findByValue](https://github.com/cerner/terra-core/blob/master/packages/terra-form-select/src/_MenuUtil.js#L125-L127)):
![2019-04-29 14 18 44](https://user-images.githubusercontent.com/34115417/56920804-cafaab00-6a89-11e9-9528-8562df81c438.gif)

1. There isn't a first class way to limit the amount of selections. This behavior is delegated to the consumer.

1. **[Minor nit]** While the `optionsFilter` prop can be used to display helper text (see example below), it may need a better description.

![2019-04-29 15 14 02](https://user-images.githubusercontent.com/34115417/56924016-8b37c180-6a91-11e9-83ce-7beaa33a7ddc.gif)

## Select2 Behavior Replication

While behavior similar to  `select2` can be replicated, it requires a lot of heavy lifting on the consumer side. Lazily loading dynamic data is a fairly common workflow/optimization when you have too many options to feasibly display on the DOM at once.

You can replicate this behavior with the current `terra-form-select` with the following logic:

```jsx
function difference(lhs, rhs) {
  return (lhs || []).filter(e => !(rhs || []).includes(e));
}

function symmetricDifference(lhs, rhs) {
  return difference(lhs, rhs)
    .concat((rhs || []).filter(e => !(lhs || []).includes(e)));
}

// maintain a list of selected values from the terra-form-select
const [selectedData, setSelectedData] = useState([]);

// memoize the content, which should be a unique & sorted list of merged selected and async data
const content = useMemo(() => {
  // get the list of nonSelected things
  const nonSelected = difference(asyncData, selectedData);

  // merge the two lists, make sure it's unique, reduceRight so that the first selected takes precedence
  // can't just use the asyncData directly because it could lack any of our currently selected values, thus running into the display bug described above
  const merged = [...selectedData, ...nonSelected]
    .reduceRight((acc, { value, display }) => ({ ...acc, [value]: display }), []);
   
  // return a sorted list of unique, merged objects 
  return Object.entries(merged)
    .map(([value, display]) => ({ display, value }))
    .sort((lhs, rhs) => lhs.display <= rhs.display ? -1 : 1);
}, [consumableData, selectedData]);

// terra-form-select onChange method
function onChange(value) {
  const diff = symmetricDifference(value, selectedValues)[0];

  if (selectedValues.includes(diff)) {
    // remove the selection if it was in the list of selectedValues
    setSelectedData(s => s.filter(({ value }) => value !== diff));
  } else {
    // add the new selection, first find the actual object in the content so we can store both the id and displayable text (value & display)
    const newSelection = content.find(({ value }) => value === diff);
    setSelectedData(s => [...s, newSelection]);
  }
}
```

## Concluding Remarks

Of all the current limitations described above, the most notable is the first.  The limitation preventing good UX for dynamic data (e.g. lazy loading data from a remote service) is the reliance on `children` options.  For this reason, we should consider making `terra-form-select` a `data-list` oriented component, rather than a `children` oriented component.

The above select2 behavior replication code snippet would be simplified if we move to a data list method. Also, the heavy lifting is delegated to the select instead of the consumer, as it should be.

---

cc: @bjankord @nickpith

**Comment by bjankord** [comment]
Related: https://github.com/cerner/terra-core/issues/2161

**Comment by StephenEsser** [comment]
I don't think switching to a data-list oriented model is necessarily the solution. The Select in it's current state can handle dynamic data fairly easily, in almost the same way as described for the Select2. 

I've put together a mock example getting as close to the first Select2 example as possible in a few minutes, and the code isn't that heavy. I'd like to compare it to the full code your gifs are demonstrating so we can narrow down the exact pain points and discuss enhancing the API to make lazy loading a little easier and more intuitive. Having a well documented example should likely be the first step.

![ezgif com-video-to-gif (2)](https://user-images.githubusercontent.com/6720991/56928826-998bda80-6a9d-11e9-886a-2122598cfb07.gif)


Source Code:

https://gist.github.com/StephenEsser/8f9515eae474b1b3eafa423485bdba21

https://gist.github.com/StephenEsser/9aee47bcfc626b3c3684449bde88737f

**Comment by dev-cprice** [comment]
@StephenEsser your example has the same `value` as `display` text

you'll see the visuals I describe when `value !== `display`

make your mock fetch return something like the following:

```jsx
function mockFetch() {
  const options = new Set();

  for (let index = 0; index < 25; index += 1) {
    options.add(Math.floor(Math.random() * 10).toString());
  }

  return [...options].map(o => ({ value: o, display: `display for ${o}`}));
}
```

**Comment by dev-cprice** [comment]
Regardless of the implementation details on how we fix this, relying on selected values to remain as children is the reason this bug exists. We need to retain a list of selected values, which includes the displayable text, not just the value.

**Comment by StephenEsser** [comment]
Interesting,

The already selected displays were flashing back to their base value, all other functionality appeared to be working as expected. 

For reference the terra-form-select follows very closely to the Ant Design API. 

https://ant.design/components/select/#components-select-demo-select-users

I'll look into seeing if their components have the same issue or if / how they've resolved it.

Ultimately I don't think we should be reinventing the wheel if this is a problem that has already been solved. We should investigate the current component libraries and determine the trade offs of each solution and what that means to the API. If those solutions don't seem to fit, we can look into our own approach. If you have a solution in mind or an example of a component you think does it well go ahead and toss that on this issue. 

Select2 is definitely something we can look into to see how they approached this issue.

**Comment by StephenEsser** [comment]
One important requirement we may need from UX on this issue before we get too far is whether or not a previously selected option is required to be within the drop down list at all times, regardless of the search criteria. I believe it's the only way a screen reader would be able to deselect an option current.

**Comment by dev-cprice** [comment]
Yeah, it would be nice if that wasn't a requirement: it would simplify things a bit.

As far as implementation goes, I was thinking something along the lines of [react-select](https://react-select.com/async)
